nssp/delphi_nssp/pull.py

```diff
@@ -1,20 +1,83 @@
 
-from .constants import (
-    NEWLINE,
-    SIGNALS,
-    SIGNALS_MAP,
-    TYPE_DICT,
-)
+from .constants import NEWLINE, SIGNALS, SIGNALS_MAP, TYPE_DICT
+
+
+def print_callback(remote_file_name, logger, bytes_so_far, bytes_total, progress_chunks):
+    """Print the callback information."""
+    rough_percent_transferred = int(100 * (bytes_so_far / bytes_total))
+    if rough_percent_transferred in progress_chunks:
+        logger.info("Transfer in progress", remote_file_name=remote_file_name, percent=rough_percent_transferred)
+        # Remove progress chunk, so it is not logged again
+        progress_chunks.remove(rough_percent_transferred)
+
+
+def get_source_data(params, logger):
+    """
+    Download historical source data from a backup server.
+
+    This function is typically used in patching only. Normal runs grab latest data from SODA API.
+
+    This function uses "user" configuration under "patch" section in params.json to specify
+    a username with local key-based access to connect to server where backup nssp source data is stored.
+    It uses "backup_dir" config under "common" section to locate backup files on remote server.
+    It then searches for CSV files that match the inclusive range of issue dates
+    specified by 'start_issue', and 'end_issue' config.
 
+    These CSV files are then downloaded and stored in the local 'source_dir' directory.
+    """
+    makedirs(params["patch"]["source_dir"], exist_ok=True)
+    ssh = paramiko.SSHClient()
+    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    host = "delphi-master-prod-01.delphi.cmu.edu"
+    user = params["patch"]["user"]
+    ssh.connect(host, username=user)
+
+    # Generate file names of source files to download
+    dates = pd.date_range(start=params["patch"]["start_issue"], end=params["patch"]["end_issue"])
+    remote_source_files = [f"{date.strftime('%Y%m%d')}.csv.gz" for date in dates]
+
+    # Download source files
+    sftp = ssh.open_sftp()
+    try:
+        sftp.stat(params["common"]["backup_dir"])
+    except IOError:
+        logger.error("Source backup directory does not exist on the remote server.")
+
+    sftp.chdir(params["common"]["backup_dir"])
+
+    num_files_transferred = 0
+    for remote_file_name in remote_source_files:
+        callback_for_filename = functools.partial(print_callback, remote_file_name, logger, progress_chunks=[0, 50])
+        local_file_path = path.join(params["patch"]["source_dir"], remote_file_name)
+        try:
+            sftp.stat(remote_file_name)
+        except IOError:
+            logger.warning(
+                "Source backup for this date does not exist on the remote server.", missing_filename=remote_file_name
+            )
+            continue
+        sftp.get(remote_file_name, local_file_path, callback=callback_for_filename)
+        logger.info("Transfer finished", remote_file_name=remote_file_name, local_file_path=local_file_path)
+        num_files_transferred += 1
+    ssh.close()
```

Comment:
suggestion (optional): You could wrap this section in a `with ssh.open_sftp() as sftp:` to avoid this `close()`. It won't change the behavior but it might make maintenance/change less error-prone.